homeassistant/components/asterisk_mbox.py

```diff
@@ -0,0 +1,171 @@
+        self.client = asteriskClient(host, port, password, self.handle_data)
+        self.messages = []
+
+        async_dispatcher_connect(
+            self.hass, SIGNAL_MESSAGE_REQUEST, self._request_messages)
+
+    @callback
+    def handle_data(self, command, msg):
+        """Handle changes to the mailbox."""
+        from asterisk_mbox.commands import CMD_MESSAGE_LIST
+
+        if command == CMD_MESSAGE_LIST:
+            _LOGGER.info("AsteriskVM sent updated message list")
+            self.messages = sorted(msg,
+                                   key=lambda item: item['info']['origtime'],
+                                   reverse=True)
+            async_dispatcher_send(self.hass, SIGNAL_MESSAGE_UPDATE,
+                                  self.messages)
+
+    @callback
+    def _request_messages(self):
+        """Handle changes to the mailbox."""
+        _LOGGER.info("Requesting message list")
+        self.client.messages()
+
+
+class AsteriskMboxMsgView(HomeAssistantView):
+    """View to return the list of messages."""
+
+    url = "/api/asteriskmbox/messages"
+    name = "api:asteriskmbox:messages"
+
+    @asyncio.coroutine
+    def get(self, request):
+        """Retrieve Asterisk messages."""
+        hass = request.app['hass']
+        msgs = hass.data[DOMAIN].messages
+        _LOGGER.info("Sending: %s", msgs)
+        return self.json(msgs)
+
+
+class AsteriskMboxDeleteView(HomeAssistantView):
+    """View to delete selected messages."""
+
+    url = "/api/asteriskmbox/delete"
+    name = "api:asteriskmbox:delete"
+
+    @asyncio.coroutine
+    def post(self, request):
+        """Delete items."""
+        try:
+            data = yield from request.json()
+
+            hass = request.app['hass']
+            client = hass.data[DOMAIN].client
+            for sha in data:
+                _LOGGER.info("Deleting: %s", sha)
+                client.delete(sha)
+        except ValueError:
+            return self.json_message('Bad item id', HTTP_BAD_REQUEST)
+
+
+class AsteriskMboxMP3View(HomeAssistantView):
+    """View to return an MP3."""
+
+    url = r"/api/asteriskmbox/mp3/{sha:[0-9a-f]+}"
+    name = "api:asteriskmbox:mp3"
+
+    @asyncio.coroutine
+    def get(self, request, sha):
+        """Retrieve Asterisk mp3."""
+        _LOGGER.info("Sending mp3 for %s", sha)
+
+        hass = request.app['hass']
+        client = hass.data[DOMAIN].client
+
+        def fetch():
+            """Read MP3 from server."""
+            from asterisk_mbox import ServerError
+
+            try:
+                return client.mp3(sha, sync=True)
+            except ServerError as err:
+                _LOGGER.error("Error getting MP3: %s", err)
+                return self.json_message(err, HTTP_NOT_FOUND)
+
+        with suppress(asyncio.CancelledError, asyncio.TimeoutError):
+            with async_timeout.timeout(10, loop=request.app['hass'].loop):
+                stream = yield from hass.async_add_job(fetch)
```

Comment:
This won't work, as you're going to return `self.json_message` if something goes wrong.

Since exceptions propagate, you can just do 

```python
from functools import partial

try:
    stream = yield from hass.async_add_job(partial(client.mp3, sha, sync=True))
except ServerError as err:
```